x-pack/plugin/monitoring/src/main/java/org/elasticsearch/xpack/monitoring/Monitoring.java

```diff
@@ -263,13 +277,71 @@ public void reload(Settings settingsToLoad) throws Exception {
     @Override
     public UnaryOperator<Map<String, IndexTemplateMetadata>> getIndexTemplateMetadataUpgrader() {
         return map -> {
+            List<IndexTemplateMetadata> monitoringTemplates = createMonitoringTemplates(getMissingMonitoringTemplateIds(map));
+            for (IndexTemplateMetadata newTemplate : monitoringTemplates) {
+                map.put(newTemplate.getName(), newTemplate);
+            }
+
+            map.entrySet().removeIf(Monitoring::isTypedAPMTemplate);
+
             // this template was not migrated to typeless due to the possibility of the old /_monitoring/bulk API being used
             // see {@link org.elasticsearch.xpack.core.monitoring.exporter.MonitoringTemplateUtils#OLD_TEMPLATE_VERSION}
             // however the bulk API is not typed (the type field is for the docs, a field inside the docs) so it's safe to remove this
             // old template and rely on the updated, typeless, .monitoring-alerts-7 template
             map.remove(".monitoring-alerts");
             return map;
         };
+    }
+
+    /**
+     * Returns a list of template IDs (as defined by {@link MonitoringTemplateUtils#TEMPLATE_IDS}) that are not present in the provided
+     * map or don't have at least {@link MonitoringTemplateUtils#LAST_UPDATED_VERSION} version
+     */
+    static List<String> getMissingMonitoringTemplateIds(Map<String, IndexTemplateMetadata> map) {
+        return Arrays.stream(MonitoringTemplateUtils.TEMPLATE_IDS).filter(id -> {
+            IndexTemplateMetadata templateMetadata = map.get(templateName(id));
+            return templateMetadata == null || (templateMetadata.version() != null && templateMetadata.version() < LAST_UPDATED_VERSION);
+        }).collect(Collectors.toList());
+    }
 
+    /**
+     * Creates the monitoring templates with the provided IDs (must be some of {@link MonitoringTemplateUtils#TEMPLATE_IDS}).
+     * Other ids are ignored.
+     */
+    static List<IndexTemplateMetadata> createMonitoringTemplates(List<String> missingTemplateIds) {
+        List<IndexTemplateMetadata> createdTemplates = new ArrayList<>(missingTemplateIds.size());
+        for (String templateId : missingTemplateIds) {
+            try {
+                final String templateName = MonitoringTemplateUtils.templateName(templateId);
+                final String templateSource = MonitoringTemplateUtils.loadTemplate(templateId);
+                try (
+                    XContentParser parser = XContentType.JSON.xContent()
+                        .createParser(NamedXContentRegistry.EMPTY, DeprecationHandler.THROW_UNSUPPORTED_OPERATION, templateSource)
+                ) {
+                    IndexTemplateMetadata updatedTemplate = IndexTemplateMetadata.Builder.fromXContent(parser, templateName);
+                    logger.info("created template [{}] with version [{}]", templateName, MonitoringTemplateUtils.TEMPLATE_VERSION);
```

Comment:
nit: s/created/creating (unless this matches other messages) Ideally the log message would happen after the actual action was performed.